streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1293,12 +1306,16 @@ private void setRegexMatchedTopicToStateStore() {
 
     private <S extends StateStore> InternalTopicConfig createChangelogTopicConfig(final StateStoreFactory<S> factory,
                                                                                   final String name) {
-        if (factory.isWindowStore()) {
+        if (factory.isVersionedStore()) {
+            final VersionedChangelogTopicConfig config = new VersionedChangelogTopicConfig(name, factory.logConfig());
+            config.setMinCompactionLagMs(factory.historyRetention());
```

Comment:
Why are we setting this "outside" but not via the constructor? (I see that we do the same thing for `WindowedChangelogTopicConfig` but I am not sure why either.)

Reply:
Do we also need to increase `delete.retention.ms` for this case? Not sure how both config "interact".

Reply:
> Why are we setting this "outside" but not via the constructor? (I see that we do the same thing for WindowedChangelogTopicConfig but I am not sure why either.)

No reason, AFAICT -- I was just following precedent. Would you prefer passing `factory.historyRetention()` into the constructor and setting it there? I can refactor `WindowedChangelogTopicConfig` as well if so.

Reply:
> Would you prefer passing factory.historyRetention() into the constructor and setting it there? I can refactor WindowedChangelogTopicConfig as well if so.

Yes, I think it would be cleaner this way.

> Nope, the changelog topic for versioned stores is only compacted, so delete.retention.ms is unused. (In contrast to the windowed changelog topic case, where both compaction and retention-based deletion are enabled)

`delete.retetion.ms` is not for retention based topics, but it's for compacted topic -- it defines how long tombstones are retained if a tombstone is the last message for a key -- ie, a tombstone marks a "soft delete" and is preserved for a while, but not directly removed when compaction runs. The goal is to ensure that a potentially offline consumer does not miss a tombstone, what would lead to a "data leak" as it might never cleanup a record.

Reply:
> Yes, I think it would be cleaner this way.

OK, made the updates.

> `delete.retetion.ms` is not for retention based topics, but it's for compacted topic

Ah I see. My scala's not the best but it looks like `min.compaction.lag.ms` guarantees that any record within `min.compaction.lag.ms` of "now" will not be compacted, regardless of `delete.retention.ms`, which is the important point that we need to guarantee that older record versions are not prematurely compacted.

Interestingly, we might have a case for setting `delete.retention.ms = 0` (rather than using the default of 24 hours) since we know that we no longer need the older tombstones once `min.compaction.lag.ms` is expired. It's not strictly necessary though. Do you think we should set it for completeness? I'm fine either way.

Reply:
>  It's not strictly necessary though.

If not necessary, no need to do anything. Just wanted to probe if we need to do anything.